linode/image/framework_resource.go

```diff
@@ -0,0 +1,414 @@
+				linodeID, diskID,
+			),
+			err.Error(),
+		)
+		return nil
+	}
+
+	createOpts := linodego.ImageCreateOptions{
+		DiskID:      diskID,
+		Label:       plan.Label.ValueString(),
+		Description: plan.Description.ValueString(),
+		CloudInit:   plan.CloudInit.ValueBool(),
+	}
+	tflog.Trace(ctx, "client.CreateImage(...)", map[string]any{
+		"options": createOpts,
+	})
+
+	image, err := client.CreateImage(ctx, createOpts)
+	if image != nil && len(image.ID) > 0 {
+		addImageResource(ctx, resp, image.ID)
+	}
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating a Linode Image", err.Error())
+		return image
+	}
+
+	ctx = populateLogAttributes(ctx, image.ID)
+	tflog.Debug(ctx, "Waiting for a single image to be ready")
+
+	tflog.Trace(ctx, "client.WaitForInstanceDiskStatus(...)")
+
+	if _, err := client.WaitForInstanceDiskStatus(
+		ctx, linodeID, diskID, linodego.DiskReady, timeoutSeconds,
+	); err != nil {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf(
+				"Failed to wait for linode instance %d disk %d to become ready while taking an image",
+				linodeID, diskID,
+			),
+			err.Error(),
+		)
+	}
+
+	// Override unknown hash to null.
+	// Hash is only known when uploading image from file
+	plan.FileHash = helper.KeepOrUpdateValue(plan.FileHash, types.StringNull(), true)
+
+	return refreshImage(ctx, image, client, &resp.Diagnostics)
+}
+
+func (r *Resource) Create(
+	ctx context.Context,
+	req resource.CreateRequest,
+	resp *resource.CreateResponse,
+) {
+	tflog.Debug(ctx, "Create "+r.Config.Name)
+
+	var plan ResourceModel
+	client := r.Meta.Client
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	createTimeout, diags := plan.Timeouts.Create(ctx, DefaultVolumeCreateTimeout)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	timeoutSeconds := helper.FrameworkSafeFloat64ToInt(createTimeout.Seconds(), &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var image *linodego.Image
+	if !plan.LinodeID.IsNull() && plan.FilePath.IsNull() {
+		image = createResourceFromLinode(ctx, &plan, client, resp, timeoutSeconds)
+	} else {
+		image = createResourceFromUpload(ctx, &plan, client, resp, timeoutSeconds)
+	}
+
+	plan.FlattenImage(image, true, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// IDs should always be overridden during creation (see #1085)
+	// TODO: Remove when Crossplane empty string ID issue is resolved
+	plan.ID = types.StringValue(image.ID)
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (r *Resource) Read(
+	ctx context.Context,
+	req resource.ReadRequest,
+	resp *resource.ReadResponse,
+) {
+	tflog.Debug(ctx, "Read "+r.Config.Name)
+
+	client := r.Meta.Client
+
+	var state ResourceModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	imageID := state.ID.ValueString()
```

Comment:
I think we'll need to call [helper.FrameworkAttemptRemoveResourceForEmptyID(...)](https://github.com/linode/terraform-provider-linode/blob/dev/linode/stackscript/framework_resource.go#L102) here 👍

Reply:
Thank you! Just added